### ISSUE TITLE
Search - App freezes after cache and cookie clean of chat filter

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#49282]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.35-7
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Found when validating PR :** https://github.com/Expensify/App/pull/49258
**Email or phone of affected tester (no customers):** applausetester+bp0916w@applause.expensifail.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
Precondition; logged in application.

1. Go to Search page > Chats
2. Click on Filters > Select some users for From field and save the search
3. Go to Troubleshoot > Reset cache and restart

## Expected Result:
App doesn't crash. App doesn'r freeze

## Actual Result:
App is freezed

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/a3ed06fc-0200-4ae0-b91c-7a0f470f6a15

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by izarutskaya** [comment]
We think this issue might be related to the #wave-control

**Comment by Christinadobrzyn** [comment]
Asking everyone in this PR to take a peek - https://github.com/Expensify/App/pull/49258

**Comment by DylanDylann** [comment]
I can't reproduce this issue on the latest main 

https://github.com/user-attachments/assets/9fbfb253-adaf-48b1-b206-37ede9848ad8

**Comment by mountiny** [comment]
I cannot reproduce in staging, it is testing feature

**Comment by mountiny** [comment]
https://github.com/user-attachments/assets/33909d59-b023-4af8-ba71-2c8d8a220b46

**Comment by mvtglobally** [comment]
Multiple testers were able to reproduce this one. Let me check now again. 
https://github.com/user-attachments/assets/0cf790f7-c925-4d9e-96d6-0cb3e74793d3

**Comment by grgia** [comment]
I can reproduce it on staging https://expensify.slack.com/archives/C01GTK53T8Q/p1726505749566979 but I agree let's not block deploy on this

**Comment by grgia** [comment]
https://expensify.slack.com/archives/C01GTK53T8Q/p1726505749566979

**Comment by mountiny** [comment]
cc @luacmartins @lakchote

**Comment by lakchote** [comment]
> cc @luacmartins @lakchote

Tried again and couldn't reproduce, @luacmartins could you test it on your side?

**Comment by luacmartins** [comment]
Yea, I could reproduce on staging. I agree that this is low priority given it requires the troubleshoot menu, but still something we should look into

https://github.com/user-attachments/assets/729d1369-60ee-43ee-b744-ecab16811458

**Comment by Christinadobrzyn** [comment]
thanks for reviewing @luacmartins - is this a regression of https://github.com/Expensify/App/pull/49258 or a different issue?

**Comment by luacmartins** [comment]
Apparently this has been an [issue for a while](https://expensify.slack.com/archives/C01GTK53T8Q/p1726551869457679?thread_ts=1726505749.566979&cid=C01GTK53T8Q), so I don't think that's a regression form the linked issue

**Comment by Christinadobrzyn** [comment]
Awesome! Thanks for confirming @luacmartins. I'll see if SWM might be able to investigate this - https://expensify.slack.com/archives/C04878MDF34/p1726679430131789

**Comment by Kicu** [comment]
hey, I will spend some time checking this and see how much work this might be. I will post any updates later.

**Comment by luacmartins** [comment]
@Kicu I would prioritize the other Search v2.4 issues over this one, since this issue is quite low value given that the troubleshoot menu is only available on dev/staging and not to customers.

**Comment by Kicu** [comment]
OK I spent a few minutes on it, I find it impossible to reproduce on my dev environment - the "clear cache and restart" for me never finishes so I cannot see what in Search makes this flow break.
Is it only clear cache after search that makes the app freeze?

**Comment by luacmartins** [comment]
I saw a comment on Slack about it happening before as well, so not entirely sure. Anyways, I think we can focus on other issues for now.

**Comment by Christinadobrzyn** [comment]
This is going to be worked on by @Guccio163 when they are available.

**Comment by Christinadobrzyn** [comment]
heads up, I'm going to be ooo for the rest of the week. I don't think there's any reason to assign a new BZ teammate as we are waiting on @Guccio163 to work on this when they are ready.

**Comment by Christinadobrzyn** [comment]
checking in with @Guccio163 on taking this - https://expensify.slack.com/archives/C04878MDF34/p1727656958314559?thread_ts=1726679430.131789&cid=C04878MDF34

**Comment by Christinadobrzyn** [comment]
checking in with @Guccio163 on taking this - https://expensify.slack.com/archives/C04878MDF34/p1727656958314559?thread_ts=1726679430.131789&cid=C04878MDF34

**Comment by Guccio163** [comment]
Hi, I'm Wiktor Gut from SWM, please assign me this issue

**Comment by Christinadobrzyn** [comment]
We're okay Melvin - this job is being worked on accordingly!

**Comment by Guccio163** [comment]
Hi @Christinadobrzyn, I'm just coming with an update on this issue; Right now I'm unable to reproduce this bug both on local and staging. After cleaning the cache one of two possible scenarios happen:
1. App refreshes gracefully:

https://github.com/user-attachments/assets/44afebe4-ba61-49e5-9a8a-93ac20a53417

2. Profile does not load, Inbox half-fully gives up on loading and search loads normally. I needed to refresh page and/or login once again to make it load:

https://github.com/user-attachments/assets/c61c2262-9891-4746-ae69-22788a839c47



I asked some guys from SWM to check if they can reproduce it to find a cause why I can't, I'll come back to you if I figure it out 🔜

**Comment by Christinadobrzyn** [comment]
Ah thank you for the update @Guccio163! Please let us know if there's anything we can do to help!

**Comment by Christinadobrzyn** [comment]
Asked QA to retest - https://expensify.slack.com/archives/C9YU7BX5M/p1728370749611919

**Comment by Guccio163** [comment]
Hi, I reached out to couple of colleagues from SWM to try and reproduce the bug, but no-one could. Nevertheless I see a big issue with `Clear cache and restart` button - It works non-deterministically sometimes right and sometimes so wrong it requires multiple cookie clean and manual refresh. I'm waiting for @izarutskaya's results though, we should decide what to do with this issue then 🔜

**Comment by kavimuru** [comment]
Able to reproduce freeze. Also user need user need to clean cache multiple times https://github.com/Expensify/App/issues/49282#issuecomment-2399147284 is true.

https://github.com/user-attachments/assets/5dd3685b-bb6b-48c0-939e-a44969c4dd30

**Comment by Christinadobrzyn** [comment]
Thanks @kavimuru - @Guccio163 will you please let us know if you need anything else to continue investigating/testing?

**Comment by Guccio163** [comment]
Hi @Christinadobrzyn, is there a specified number of refreshes needed for it to break? So far I got up to ~20 and I still cannot reproduce this one.

https://github.com/user-attachments/assets/93cc7ee7-ab88-4935-850e-7c99d056a310

**Comment by Christinadobrzyn** [comment]
Asking @kavimuru to review the questions - https://expensify.slack.com/archives/C9YU7BX5M/p1728624548081019?thread_ts=1728370749.611919&cid=C9YU7BX5M

**Comment by kavimuru** [comment]
@Guccio163 @Christinadobrzyn App freezes for me right after " Reset cache and restart" @

https://github.com/user-attachments/assets/cc555def-4714-46a5-a225-43a63afc0c94

**Comment by tylerkaraszewski** [comment]
People are working on it melvin.

**Comment by Christinadobrzyn** [comment]
Feel free to let us know how we can help with testing this https://github.com/Expensify/App/issues/49282#issuecomment-2407603065 - TY!

**Comment by Guccio163** [comment]
Hi @kavimuru, could you please attach some technical specification of the device you're reproducing on? I've noticed that's MacBook/iMac, but I'm working on one too, so maybe it's the specification differences.

**Comment by Guccio163** [comment]
Thanks for the insight @rayane-djouah, I'll try this one in a moment! 🫶

**Comment by Guccio163** [comment]
Hi @Christinadobrzyn, I'm coming with my conclusions on this issue: There's a list of tests I performed here to try and reproduce the bug: 

Reported bug behaviour:
App freezes and doesn't respond

- Tried to follow reproduction steps - the bug didn't happen
- Tried  to spot the bug right after the cache clear as per [this](https://github.com/Expensify/App/issues/49282#issuecomment-2407603065) comment - the bug didn't happen 
- Tried refreshing multiple times as per [this](https://github.com/Expensify/App/issues/49282#issuecomment-2399869480) comment - the bug didn't happen
- Tried reproducing on a smaller window as per [this](https://github.com/Expensify/App/issues/49282#issuecomment-2413367734) comment (visible below) - the bug didn't happen

Current behaviour (One of two, non-deterministic) :
1. User's details and inbox loading infinitely
2. All informations loaded right after refresh

Moreover I'm pretty much sure that this bug isn't connected to Search at all; To illustrate the non-deterministic nature of this bug, below are recordings of app behaviour with and without the search-related steps:
1. With selecting search filters
2. Without selecting filters (just clearing cache)

### Please notice that no matter with or without filters, app behaves the same in one of 2 defined behaviour above: ###
<details>
<summary>Staging: filters then cache clear - loading loop </summary>
https://github.com/user-attachments/assets/c68ca794-25e0-4904-bf35-d8ff7b5b2f55
</details>
<details>
<summary>Staging: filters then cache clear - loading normally </summary>
https://github.com/user-attachments/assets/af9b4aab-2fa7-4376-98bd-61d2c2bf8bec
</details>
<details>
<summary>Staging: Just cache clear - loading loop </summary>
https://github.com/user-attachments/assets/fb362bdd-4d20-40c6-861b-0b625173b5dd
</details>
<details>
<summary>Staging: Just cache clear - loading normally </summary>
https://github.com/user-attachments/assets/bd3a3ee7-1d3a-4ade-beab-58f1e2ed04da
</details>

### Also, all these results are possible to spot in Local environment too: ###

<details>
<summary> Local: filters then cache clear - loading loop </summary>
https://github.com/user-attachments/assets/4647ada4-a39a-413b-8e93-218dbb4e0620
</details>
<details>
<summary> Local: filters then cache clear - loading normally </summary>
https://github.com/user-attachments/assets/55fa3a97-cbb1-4e37-a3cd-6a23af79904c
</details>
<details>
<summary> Local: Just cache clear - loading loop </summary>
https://github.com/user-attachments/assets/20bd7dde-ce3b-4b79-a66b-8f1874effe4f
</details>
<details>
<summary>Local: Just cache clear - loading normally </summary>
https://github.com/user-attachments/assets/8a566938-127c-41ac-85bd-56156aec2beb
</details>

### As a final addition, exactly the same infinite loading happens (non-deterministic) right after logging in: ###
<details>
<summary>Staging: Right after logging in - loading loop </summary>
https://github.com/user-attachments/assets/ce33ac67-da91-4071-9d3b-9dc144872d7d
</details>

In my opinion incorrect behaviour isn't linked to Search and is purely matter of "Clear cache and restart" button; Maybe the contributor that added this feature or someone with bigger Onyx knowledge could take this and dive into the button's functionality?

**Comment by Christinadobrzyn** [comment]
Hi @Guccio163! Thank you so much for the thorough review. It sounds like we need to find an Onyx person to investigate more. I think @kidroca does Onyx stuff but I don't know if it relates to this kind of job. 

@tylerkaraszewski or @rayane-djouah do you know who else we might be able to reach out to about this?

**Comment by Christinadobrzyn** [comment]
Awesome, thanks @rayane-djouah! I reached out to the Callstack team to see if they might have someone with Onyx experience to help! https://expensify.slack.com/archives/C03UK30EA1Z/p1729654603736149

**Comment by TMisiukiewicz** [comment]
Hey, I'll take a look on this later this week 👍

**Comment by Christinadobrzyn** [comment]
Awesome! Thank you @TMisiukiewicz!

**Comment by Christinadobrzyn** [comment]
Thanks for the testing @TMisiukiewicz - reaching out to QA to see if they can test again based on those instructions. https://expensify.slack.com/archives/C9YU7BX5M/p1730176431155529

**Comment by izarutskaya** [comment]
Still reproducible but I am not sure what should I see. Check please my video
Build v9.0.55-2
https://github.com/user-attachments/assets/581fa8aa-6516-4df6-8424-e5f3663ded7e

**Comment by TMisiukiewicz** [comment]
@izarutskaya thanks, looks like it hangs exactly at the same point.

So today I was finally able to reproduce it and found the root cause. When we open the Search in chats selector, we initialize the option list to load the items that are visible in the list:
https://github.com/Expensify/App/blob/65fa00a002b3603d335e9a3180b10d6fa365cd58/src/components/Search/SearchFiltersChatsSelector.tsx#L46-L48

This means, now every time the `personalDetails` list change, we will fall into this useEffect and execute the whole code since the early return won't work anymore:
https://github.com/Expensify/App/blob/65fa00a002b3603d335e9a3180b10d6fa365cd58/src/components/OptionListContextProvider.tsx#L82-L128

As you can see, it loops through all the personal details, and for each single one of them, it loops through all reports using `filter` and then `forEach`. This issue probably occurs only on heavy loaded accounts - for me it's reproducible on account where I have 8,5k personal details and more than 15k reports. I was able to measure it using profiler and it blocks my app for 50 seconds:

![image](https://github.com/user-attachments/assets/9f20a350-4388-46cc-ab89-132e5bbccbcd)

And here is the confirmation that loops mentioned earlier are responsible for that:

![image](https://github.com/user-attachments/assets/ac149a82-2829-4200-9bab-c37ba8ac4103)

In this useEffect we have early return comparing the previous value of personal details with the current one, and it should do the work in all the other cases where this value gets updated. However in here it's kinda specific scenario, and it will never fall into this early return because the previous value of personal details is just an empty object, so we run the code for each one of them.

Haven't tested any solution yet, but first and easiest thing that comes to my mind is just setting `areOptionsInitialized` to `false` when clearing the cache. Will test it tomorrow morning and open a PR if it works properly.

**Comment by TMisiukiewicz** [comment]
PR open: https://github.com/Expensify/App/pull/51725

mind checking if the issue is gone on your side when testing out this branch?

**Comment by TMisiukiewicz** [comment]
I'll be ooo until the end of the week so @MrRefactor will take care of this if needed

**Comment by MrRefactor** [comment]
I will adjust PR and upload all of the recordings necessary!

**Comment by Christinadobrzyn** [comment]
Update for Melvin Monitoring PR - https://github.com/Expensify/App/pull/51725

**Comment by Christinadobrzyn** [comment]
moving to daily since the payment date is coming 

Payment summary here - https://github.com/Expensify/App/issues/49282#issuecomment-2472038608

@rayane-djouah Do we need a regression test for this?

**Comment by Christinadobrzyn** [comment]
moving to daily since the payment date is coming 

Payment summary here - https://github.com/Expensify/App/issues/49282#issuecomment-2472038608

@rayane-djouah Do we need a regression test for this?

**Comment by Christinadobrzyn** [comment]
Payment day - no regressions, 

Payouts due:

- [X] Reviewer: $250 @rayane-djouah (Paid in upwork - https://www.upwork.com/nx/wm/offer/104197919)

Closing this out as complete
